L1Trigger/TrackFindingTracklet/plugins/ProducerKFin.cc

```diff
@@ -0,0 +1,261 @@
+          double phiT = deltaPhi(ttTrackRef->phi() - setup_->hybridChosenRofPhi() * qOverPt - ttTrackRef->phiSector() * setup_->baseRegion());
+          const int sectorPhi = phiT < 0. ? 0 : 1; // dirty hack
+          phiT -= (sectorPhi - .5) * setup_->baseSector();
+          // cut on nonant size and pt
+          if (!dfPhiT.inRange(phiT) || !dfQoverPt.inRange(qOverPt))
+            continue;
+          // loop over stubs
+          TTBV hitPattern(0, setup_->numLayers());
+          vector<int> layerMap(setup_->numLayers(), 0);
+          for (const TTStubRef& ttStubRef : ttTrackRef->getStubRefs()) {
+            const GlobalPoint& gp = setup_->stubPos(ttStubRef);
+            const double r = gp.perp() - setup_->hybridChosenRofPhi();
+            const double phi = deltaPhi(gp.phi() - (ttTrackRef->phi() - qOverPt * gp.perp()));
+            const double z = gp.z() - (ttTrackRef->z0() + ttTrackRef->tanL() * gp.perp());
+            double layer = distance(layerEncoding.begin(), find(layerEncoding.begin(), layerEncoding.end(), setup_->layerId(ttStubRef)));
+            if (layer >= setup_->numLayers())
+              layer = setup_->numLayers() - 1;
+            // cut on phi and z residuals
+            if (!dfPhi.inRange(phi) || !dfZ.inRange(z))
+              continue;
+            hitPattern.set(layer);
+            int& nLayerStubs = layerMap[layer];
+            // cut on max 4 stubs per layer
+            if (nLayerStubs == setup_->kfMaxStubsPerLayer())
+              continue;
+            nLayerStubs++;
+            numLayerStubs[layer]++;
+            stubs.emplace_back(ttStubRef, dataFormats_, r, phi, z, trackId, layer);
+          }
+          const TTBV& maybePattern = layerEncoding_->maybePattern(binEta, binZT, binCot);
+          tracks.emplace_back(ttTrackRef, dataFormats_, hitPattern, setup_->layerMap(hitPattern, layerMap), maybePattern, phiT, qOverPt, zT, cot, sectorPhi, binEta, trackId++);
+        }
+        // truncate and transform stubs
+        vector<vector<StubKFin*>> layerStubs(setup_->numLayers());
+        for (int layer = 0; layer < setup_->numLayers(); layer++)
+          layerStubs[layer].reserve(numLayerStubs[layer]);
+        for (StubKFin& stub : stubs)
+          layerStubs[stub.layer()].push_back(&stub);
+        const int offset = channelId * setup_->numLayers();
+        for (int layer = 0; layer < setup_->numLayers(); layer++) {
+          TTDTC::Stream& acceptedStubs = streamAcceptedStubs[offset + layer];
+          TTDTC::Stream& lostStubs = streamLostStubs[offset + layer];
+          vector<StubKFin*>& stubsLayer = layerStubs[layer];
+          if (enableTruncation_ && (int)stubsLayer.size() > setup_->numFrames()) {
+            const auto it = next(stubsLayer.begin(), setup_->numFrames());
+            lostStubs.reserve(setup_->numFrames() - (int)stubsLayer.size());
+            transform(it, stubsLayer.end(), back_inserter(lostStubs), toFrameStub);
```

Comment:
If truncation could be applied using similar code to used in TrackFindingTFP/plugins/ProducerKFin.cc, it would simplify things. (Understand one = understand both). I think this would be possible if vector layerStubs stored stub->frame() instead of stub.